src/record_types.jl

```diff
@@ -12,10 +12,10 @@ Allocation record stores all the allocations created by `Array`. You can use
 [`record_allocations`](@ref) to create such a record.
 """
 struct AllocationRecord <: AbstractAllocationRecord
-    allocations::Vector{Array}
+    allocations::Vector{AbstractArray}
```

Comment:
This does not seem to create any inference issues, and I added several tests for this.

Reply:
This should cause massive issues, since the compiler can't know what subtype is coming out when it does `getindex`.
I would like to know why it doesn't.
Is it because we are putting in type assertions based on the constructor when we are doing the replay?

Reply:
That's my going understanding, yea.

Reply:
lets add a comment to that effect

Reply:
Will do. Should I simplify and get rid of the whole "freezing" machinery too? Since I don't think it can really help with inference beyond that type assert?

Reply:
> Will do. Should I simplify and get rid of the whole "freezing" machinery too? Since I don't think it can really help with inference beyond that type assert?

I don't fully recall what the `freezing` machinery was for,
but presumably changing it is out of scope for this PR?